### ISSUE TITLE
DCQL: Define details for the purpose property

### ISSUE THREAD

**Issue by danielfett** [openid/OpenID4VP#289]
As of now, the purpose property is completely undefined. We should define a structure for it.

Related to #160

**Comment by bc-pi** [comment]
Seems related to #230 too

**Comment by martijnharing** [comment]
Can we make this 1.0? It's an open discussion point to remove the option to have strings as purpose because of security reasons.

**Comment by Sakurann** [comment]
we discussed this one among co-chairs and editors and the current thinking has been that we can tackle this in 1.1 without introducing breaking changes. this is an optional parameter, so we can later add a better mechanism and deprecate this one. or add a new parameter on top of the current parameter. Based on the discussion in the WG so far, it seems extremely unlikely we will be able to agree on this "better solution" before 1.0.

**Comment by martijnharing** [comment]
The current definition in the specification allows freely defined text strings, something that has some significant privacy and usability problems, some examples:
- Since the wallet does not understand these text strings, it cannot change its UI based on the type of purpose. Making it less likely to be able to properly help the user in making its choices.
- Localization is difficult and/or cumbersome to deal with.
- It makes it harder to design for accessibility
- It would require the wallet to show text that it doesn't understand in UI that can be considered trustworthy by the user. Leading to a range of possible injection attacks.

While we figure as an ecosystem how to best solve this, it's important that we take a conservative approach with respect to this feature. This ensures that we protect the user and direct RPs, issuers etc to work towards a solution that achieves that as well.
At this point, having some form of enum that can be ecosystem specific (to allow ecosystems to come up with their own appropriate values while ensuring non-collision) is a good place to start.

One (compromise) solution could be to say that for 1.0 free-text strings are not allowed and must be ignored by the wallet, but allow the structure to support that. This does require a change to the 1.0 specification, which is why this issue should be labeled 1.0.

**Comment by charsleysa** [comment]
There are already implementations supporting purpose as a free text short title field so this would need to be taken into consideration for structure changes. I would not support complete removal without a viable alternative.

**Comment by Sakurann** [comment]
WG discussion (trying my best to summarize):
- proposal to remove purpose parameter on the grounds of:
    - human injection attacks (@bc-pi can you please elaborate?) (agreed technical injection attacks are solvable)
    - purpose property is not useful/meaningful to the user?
    - remove now, discuss more and re-add in 1.1
- objections to remove this parameter on the grounds of:
    - alternative solutions are not better than the current parameter
        - alternative solutions being... an ecosystem specific list of purpose codes.
    - but how to enforce ecosystem specific code when in many cases RP does not know the ecosystem the wallet belongs to... (why we added multi-RP request)

**Comment by bc-pi** [comment]
> * human injection attacks ([@bc-pi](https://github.com/bc-pi) can you please elaborate?)

Where a bad acting verifier is able to have alarming or misleading text displayed to the user in the context of the wallet/browser/OS UI, which has a different/higher level of trust to the user. 

My suggestion/proposal is that the verifier display purpose or whatever necessary context in it's own UI before or around the 4VP interaction. And the wallet/browser/OS UI display factually what will be revealed based on the DQCL content. 

This doesn't stop a bad acting verifier from saying misleading things but does keep it from saying them in the context of a more/differently trusted component.

**Comment by Sakurann** [comment]
WG discussion:
- if purpose is in a signed request, less concerns for human injection attacks? is it a compromise here to say `purpose` may be present only in the signed request? being in the signed request does not mean "trusted"?
- another proposal: put purpose in KB-JWT/deviceSigned
- Brian's suggestion above is the most viable among those that have been proposed?
- rough agreement that integers/pre-defined lists do not solve the problem..
- **larger preference seems to be to remove this from 1.0 Final for now and elaborate on Brian's suggestion above in the specification, especially on the UI challenges**
- IIW session will solve everything :)

**Comment by c2bo** [comment]
Yeah, it might make sense to introduce an ID `credential_sets` so they can be referenced easier, especially if that is what we envision for purpose.

**Comment by Sakurann** [comment]
btw ARF Annex 2 RPA_09 says
"A Relying Party SHOULD communicate in the request which attributes are needed for which purpose (use case, service), if this is supported by the protocol used for communication with the Wallet Unit."`
